File: ios/LearnWithAlphonsoKit/Sources/LearnWithAlphonsoKit/SpokenAnswer.swift
import Foundation

/// Comparing a speech-to-text transcript against the phrase the learner was
/// asked to say.
///
/// A hand-kept port of src/lib/spoken-answer.ts, which also has a Deno copy in
/// supabase/functions/grade-review. Three copies is not a happy number, but the
/// rule has to hold in three runtimes: this app grades optimistically for
/// immediate feedback, and the server re-derives the same verdict afterwards.
/// If they disagree the learner is congratulated and then has the item lapsed
/// behind their back, which is invisible from either side alone. The vectors in
/// SpokenAnswerTests are the same vectors as the TypeScript and Deno tests, and
/// that is what keeps the three honest.
///
/// Deliberately NOT edit distance: a threshold loose enough to forgive
/// "she is"/"she's" also accepts "he is a driver" for "she is a doctor".
enum SpokenAnswer {
    /// Expanded rather than contracted, so "don't" and "do not" converge.
    private static let contractions: [(String, String)] = [
        ("\\bcan't\\b", "can not"),
        ("\\bwon't\\b", "will not"),
        ("\\bn't\\b", " not"),
        ("\\b'll\\b", " will"),
        ("\\b're\\b", " are"),
        ("\\b've\\b", " have"),
        ("\\b'd\\b", " would"),
        // Possessive and "is" share this form; spoken practice uses it as "is",
        // and treating a possessive as "is" only ever makes two spellings of
        // the same utterance agree.
        ("\\b's\\b", " is"),
        ("\\b'm\\b", " am"),
    ]

    /// The same contractions WITHOUT an apostrophe, which is how speech
    /// recognition frequently returns them ("shes a doctor"). Every key is a
    /// non-word in English deliberately: expanding "were" or "well" would
    /// corrupt real words, so those forms are left alone. Under-matching costs
    /// a retry; over-matching marks a wrong answer right.
    private static let apostropheLess: [(String, String)] = [
        ("\\bcant\\b", "can not"),
        ("\\bwont\\b", "will not"),
        ("\\bdont\\b", "do not"),
        ("\\bdoesnt\\b", "does not"),
        ("\\bdidnt\\b", "did not"),
        ("\\bisnt\\b", "is not"),
        ("\\barent\\b", "are not"),
        ("\\bwasnt\\b", "was not"),
        ("\\bwerent\\b", "were not"),
        ("\\bhasnt\\b", "has not"),
        ("\\bhavent\\b", "have not"),
        ("\\bhadnt\\b", "had not"),
        ("\\bcouldnt\\b", "could not"),
        ("\\bwouldnt\\b", "would not"),
        ("\\bshouldnt\\b", "should not"),
        ("\\bshes\\b", "she is"),
        ("\\bhes\\b", "he is"),
        ("\\btheres\\b", "there is"),
        ("\\bthats\\b", "that is"),
        ("\\bwhats\\b", "what is"),
        ("\\blets\\b", "let us"),
        ("\\bim\\b", "i am"),
        ("\\bive\\b", "i have"),
        ("\\byoure\\b", "you are"),
        ("\\btheyre\\b", "they are"),
        ("\\byouve\\b", "you have"),
        ("\\bweve\\b", "we have"),
    ]

    /// Hesitation noises the recogniser transcribes but nobody means to say.
    private static let filler = "\\b(?:um|uh|erm|er|ah)\\b"

    /// Number words collapsed onto digits, because /api/stt calls Deepgram with
    /// `smart_format=true`, which returns spoken numbers as numerals: say "the
    /// bus leaves at nine" and the transcript reads "the bus leaves at 9".
    /// Single words only -- compound numbers cannot be reconciled this way, so
    /// spoken content avoids them.
    private static let numberWords: [(String, String)] = [
        ("\\bzero\\b", "0"), ("\\bone\\b", "1"), ("\\btwo\\b", "2"), ("\\bthree\\b", "3"),
        ("\\bfour\\b", "4"), ("\\bfive\\b", "5"), ("\\bsix\\b", "6"), ("\\bseven\\b", "7"),
        ("\\beight\\b", "8"), ("\\bnine\\b", "9"), ("\\bten\\b", "10"), ("\\beleven\\b", "11"),
        ("\\btwelve\\b", "12"), ("\\bthirteen\\b", "13"), ("\\bfourteen\\b", "14"),
        ("\\bfifteen\\b", "15"), ("\\bsixteen\\b", "16"), ("\\bseventeen\\b", "17"),
        ("\\beighteen\\b", "18"), ("\\bnineteen\\b", "19"), ("\\btwenty\\b", "20"),
        ("\\bthirty\\b", "30"), ("\\bforty\\b", "40"), ("\\bfifty\\b", "50"),
        ("\\bsixty\\b", "60"), ("\\bseventy\\b", "70"), ("\\beighty\\b", "80"),
        ("\\bninety\\b", "90"),
    ]

    private static func replacing(_ input: String, _ pattern: String, with replacement: String)
        -> String
    {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return input }
        return regex.stringByReplacingMatches(
            in: input,
            range: NSRange(input.startIndex..., in: input),
            withTemplate: replacement
        )
    }

    static func normalise(_ input: String) -> String {
        var s = input.lowercased()
        // Expand while the apostrophes are still present...
        for (pattern, replacement) in contractions {
            s = replacing(s, pattern, with: replacement)
        }
        s = replacing(s, "['\u{2019}]", with: "")
        // ...then again for the apostrophe-less spellings recognition emits.
        for (pattern, replacement) in apostropheLess {
            s = replacing(s, pattern, with: replacement)
        }
        s = replacing(s, filler, with: " ")
        for (pattern, replacement) in numberWords {
            s = replacing(s, pattern, with: replacement)
        }
        s = replacing(s, "[^a-z0-9\\s]", with: " ")
        s = replacing(s, "\\s+", with: " ")
        return s.trimmingCharacters(in: .whitespaces)
    }

    /// Whether `transcript` is the learner saying `expected`.
    ///
    /// An empty transcript is false -- but it means "nothing was captured",
    /// not "said it wrong", and callers must not spend a heart on it. Recording
    /// failures reach here looking identical to silence, which is why the
    /// speaking UI never submits one.
    static func matches(transcript: String, expected: String) -> Bool {
        let said = normalise(transcript)
        if said.isEmpty { return false }
        return said == normalise(expected)
    }
}

File: supabase/functions/grade-review/spoken-answer.ts
// Deno copy of src/lib/spoken-answer.ts. Supabase Edge Functions bundle each
// function directory independently, so a relative import reaching outside
// supabase/functions/grade-review/ is not reliably resolvable -- same reasoning
// as srs.ts here and complete-lesson/progress-math.ts.
//
// KEEP IN SYNC with the source of truth. This is not cosmetic duplication: the
// player grades a spoken answer with the original and shows the learner a
// verdict, and this copy then re-derives that verdict server-side. Any drift
// means the learner is told they were right and has the item lapsed anyway.
// spoken-answer.test.ts here mirrors the source's vectors and CI's deno-tests
// job fails on divergence.

const CONTRACTIONS: [RegExp, string][] = [
  [/\bcan't\b/g, "can not"],
  [/\bwon't\b/g, "will not"],
  [/\bn't\b/g, " not"],
  [/\b'll\b/g, " will"],
  [/\b're\b/g, " are"],
  [/\b've\b/g, " have"],
  [/\b'd\b/g, " would"],
  [/\b's\b/g, " is"],
  [/\b'm\b/g, " am"],
];

/**
 * The same contractions written WITHOUT an apostrophe, which is how speech
 * recognition frequently returns them. Every key is a non-word in English on
 * purpose -- expanding "were", "well" or "ill" would corrupt real words.
 */
const APOSTROPHE_LESS: [RegExp, string][] = [
  [/\bcant\b/g, "can not"],
  [/\bwont\b/g, "will not"],
  [/\bdont\b/g, "do not"],
  [/\bdoesnt\b/g, "does not"],
  [/\bdidnt\b/g, "did not"],
  [/\bisnt\b/g, "is not"],
  [/\barent\b/g, "are not"],
  [/\bwasnt\b/g, "was not"],
  [/\bwerent\b/g, "were not"],
  [/\bhasnt\b/g, "has not"],
  [/\bhavent\b/g, "have not"],
  [/\bhadnt\b/g, "had not"],
  [/\bcouldnt\b/g, "could not"],
  [/\bwouldnt\b/g, "would not"],
  [/\bshouldnt\b/g, "should not"],
  [/\bshes\b/g, "she is"],
  [/\bhes\b/g, "he is"],
  [/\btheres\b/g, "there is"],
  [/\bthats\b/g, "that is"],
  [/\bwhats\b/g, "what is"],
  [/\blets\b/g, "let us"],
  [/\bim\b/g, "i am"],
  [/\bive\b/g, "i have"],
  [/\byoure\b/g, "you are"],
  [/\btheyre\b/g, "they are"],
  [/\byouve\b/g, "you have"],
  [/\bweve\b/g, "we have"],
];

const FILLER = /\b(?:um|uh|erm|er|ah)\b/g;

/**
 * Number words collapsed onto digits, because /api/stt calls Deepgram with
 * `smart_format=true`, which returns spoken numbers as numerals: say "the bus
 * leaves at nine" and the transcript reads "the bus leaves at 9". Without this
 * the two spellings of the same utterance never match and a learner who said
 * the phrase perfectly is marked wrong.
 *
 * Single words only. Compound numbers cannot be reconciled this way --
 * "twenty one" is two words here and "21" is one in the transcript -- so
 * spoken content avoids them rather than pretending this handles them.
 */
const NUMBER_WORDS: [RegExp, string][] = [
  [/\bzero\b/g, "0"],
  [/\bone\b/g, "1"],
  [/\btwo\b/g, "2"],
  [/\bthree\b/g, "3"],
  [/\bfour\b/g, "4"],
  [/\bfive\b/g, "5"],
  [/\bsix\b/g, "6"],
  [/\bseven\b/g, "7"],
  [/\beight\b/g, "8"],
  [/\bnine\b/g, "9"],
  [/\bten\b/g, "10"],
  [/\beleven\b/g, "11"],
  [/\btwelve\b/g, "12"],
  [/\bthirteen\b/g, "13"],
  [/\bfourteen\b/g, "14"],
  [/\bfifteen\b/g, "15"],
  [/\bsixteen\b/g, "16"],
  [/\bseventeen\b/g, "17"],
  [/\beighteen\b/g, "18"],
  [/\bnineteen\b/g, "19"],
  [/\btwenty\b/g, "20"],
  [/\bthirty\b/g, "30"],
  [/\bforty\b/g, "40"],
  [/\bfifty\b/g, "50"],
  [/\bsixty\b/g, "60"],
  [/\bseventy\b/g, "70"],
  [/\beighty\b/g, "80"],
  [/\bninety\b/g, "90"],
];

export function normaliseSpoken(input: string): string {
  let s = input.toLowerCase();
  for (const [pattern, replacement] of CONTRACTIONS) s = s.replace(pattern, replacement);
  s = s.replace(/['’]/g, "");
  for (const [pattern, replacement] of APOSTROPHE_LESS) s = s.replace(pattern, replacement);
  s = s.replace(FILLER, " ");
  for (const [pattern, replacement] of NUMBER_WORDS) s = s.replace(pattern, replacement);
  s = s.replace(/[^a-z0-9\s]/g, " ");
  return s.replace(/\s+/g, " ").trim();
}

export function matchesSpokenAnswer(transcript: string, expected: string): boolean {
  const said = normaliseSpoken(transcript);
  if (!said) return false;
  return said === normaliseSpoken(expected);
}
